fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,822 @@
+# Copyright (c) Fairlearn contributors.
+# Licensed under the MIT License.
+
+from math import ceil
+from time import time
+from ._constants import (
+    _IMPORT_ERROR_MESSAGE,
+    _KWARG_ERROR_MESSAGE,
+    _PREDICTION_FUNCTION_AMBIGUOUS,
+    _PROGRESS_UPDATE,
+    _CALLBACK_RETURNS_ERROR,
+)
+from ._backend_engine import BackendEngine
+from ._pytorch_engine import PytorchEngine
+from ._tensorflow_engine import TensorflowEngine
+from ._preprocessor import (
+    FloatTransformer,
+    _get_type,
+)
+from sklearn.base import (
+    ClassifierMixin,
+    RegressorMixin,
+    BaseEstimator,
+    TransformerMixin,
+)
+from sklearn.utils.validation import (
+    check_is_fitted,
+    check_random_state,
+    check_array,
+)
+from sklearn.exceptions import NotFittedError
+from numpy import zeros, argmax, arange
+
+
+class AdversarialFairness(BaseEstimator):
+    r"""Train complex predictors while mitigating biases in PyTorch or Tensorflow.
+
+    This model implements the supervised learning method proposed in
+    `"Mitigating Unwanted Biases with Adversarial Learning"
+    <https://dl.acm.org/doi/pdf/10.1145/3278721.3278779>`_ [1]_.
+    This algorithm takes as input two neural network
+    models, a predictor and an adversarial, defined either as a `PyTorch module
+    <https://pytorch.org/docs/stable/generated/torch.nn.Module.html>`_ or
+    `Tensorflow2 model
+    <https://www.tensorflow.org/api_docs/python/tf/keras/Model>`_. You train this
+    predictor using an API that is similar to estimators in `sklearn`.
+
+    As per [1]_, the neural network models :code:`predictor_model` and
```

Comment:
The use of 'may' in this paragraph is confusing me. I think that it's saying that the two models have to have outputs which are class probabilities, rather than a single prediction?